lib/location/method_wifi.c

```diff
@@ -0,0 +1,307 @@
+			goto return_error;
+		}
+
+		if (latest_scan_result_count > 1) {
+			/* Fill scanning results: */
+			request.wifi_scanning_result_count = latest_scan_result_count;
+
+			/* Calculate remaining time as a REST timeout (in mseconds): */
+			request.timeout_ms = ((wifi_config.timeout * 1000) -
+					   (k_uptime_get() - starting_uptime_ms));
+			if (request.timeout_ms < REST_WIFI_MIN_TIMEOUT_MS) {
+				if (request.timeout_ms < 0) {
+					/* No remaining time at all */
+					LOG_WRN("No remaining time left for requesting a position");
+					goto return_error;
+				}
+				request.timeout_ms = REST_WIFI_MIN_TIMEOUT_MS;
+			}
+
+			for (int i = 0; i < latest_scan_result_count; i++) {
+				strcpy(request.scanning_results[i].mac_addr_str,
+				       latest_scan_results[i].mac_addr_str);
+				strcpy(request.scanning_results[i].ssid_str,
+				       latest_scan_results[i].ssid_str);
+				request.scanning_results[i].channel =
+					latest_scan_results[i].channel;
+				request.scanning_results[i].rssi =
+					latest_scan_results[i].rssi;
+			}
+
+			err = rest_services_wifi_location_get(wifi_config.service, &request,
+							      &result);
+			if (err) {
+				LOG_ERR("Failed to acquire a location by using "
+					"WiFi positioning, err: %d",
+						err);
+				err = -1;
+			} else {
+				location_result.latitude = result.latitude;
+				location_result.longitude = result.longitude;
+				location_result.accuracy = result.accuracy;
+				if (running) {
+					loc_core_event_cb(&location_result);
+					running = false;
+				}
+			}
+		} else {
+			if (latest_scan_result_count == 1) {
+				/* Following statement seems to be true at least with Here
+				 * (400: bad request) and also with Skyhook (404: not found).
+				 * Thus, fail faster in this case and save the data transfer costs.
+				 */
+				LOG_WRN("Retrieving a location based on a single WiFi access point "
+					"is not possible");
+			} else {
+				LOG_WRN("No WiFi scanning results");
+			}
+			err = -1;
+		}
+	}
+return_error:
+	if (err) {
+		loc_core_event_cb_error();
+		running = false;
+	}
+}
+
+/******************************************************************************/
+
+int method_wifi_cancel(void)
+{
+	running = false;
+
+	(void)k_work_cancel(&method_wifi_start_work.work_item);
+	return 0;
+}
+
+int method_wifi_location_get(const struct loc_method_config *config)
+{
+	const struct loc_wifi_config wifi_config = config->wifi;
+
+	if (running) {
+		LOG_ERR("Previous operation ongoing.");
+		return -EBUSY;
+	}
+
+	k_work_init(&method_wifi_start_work.work_item, method_wifi_positioning_work_fn);
+	method_wifi_start_work.wifi_config = wifi_config;
+	method_wifi_start_work.starting_uptime_ms = k_uptime_get();
+	k_work_submit_to_queue(loc_core_work_queue_get(), &method_wifi_start_work.work_item);
+
+	running = true;
+
+	return 0;
+}
+
+int method_wifi_init(void)
+{
+	running = false;
+	current_scan_result_count = 0;
+	latest_scan_result_count = 0;
+
+	wifi_iface = net_if_get_default();
```

Comment:
Yes, that is a problem, that's why having some extra checking but even with that, might not be 100% proof.

>net_if_get_first_by_type(&NET_L2_GET_NAME(WIFI_MGMT))

Unfortunately that does not work. Seems that wifi is not supported for that. @rlubos, do you have any idea how to get wifi iface for sure?

Reply:
Yeah, `net_if_get_first_by_type()` won't work in this case as it looks up the interface by L2, and since ESP is offloaded device it does not define any L2.

I think `net_if_lookup_by_dev()` could be used here, the devce could be obtained with `device_get_binding(DT_LABEL(DT_NODELABEL(esp8266)))` - I hope I did not mix the DT macrology here.

Reply:
Thanks @rlubos. But I think that we want to be generic at this level and thus cannot do anything that is esp specific here.

Reply:
The Wi-Fi device label could be a Kconfig option?

Reply:
I think we're quite limited by upstream in this regard. I'd expect there to be a generic node, something like `zephyr,wifi` that you would then set to appropriate device in the DT overlay file, such label, however, doesn't seem to exist yet.

A Kconfig option allowing to specify device label might be a viable workaround for this, as suggested.

Reply:
Kconfig is a good idea for this one.

Reply:
Changed to be based on Kconfig + device_get_binding() + net_if_lookup_by_dev().